### PR TITLE
[clang-format] Add option to remove leading blank lines

### PR DESCRIPTION
**Pull request by sstwcw**

### CODE REVIEW

**Comment by sstwcw** [comment]
Should I make dropping leading blank lines the default?

There are some tests verifying that files that start with newlines should still start with newlines after being formatted.  However, they seem to be there to prevent the newlines from moving to wrong places.

There is 303baf53b1 "skip empty lines and comments in the top of the code when inserting new headers".

```C++
  std::string Code = "\nint x;";
  std::string Expected = "\n#include \"fix.h\"\n"
                         "#include \"a.h\"\n"
                         "#include \"b.h\"\n"
                         "#include \"c.h\"\n"
                         "#include <list>\n"
                         "#include <vector>\n"
                         "int x;";
  tooling::Replacements Replaces = toReplacements(
      {createInsertion("#include \"a.h\""), createInsertion("#include \"c.h\""),
       createInsertion("#include \"b.h\""),
       createInsertion("#include <vector>"), createInsertion("#include <list>"),
       createInsertion("#include \"fix.h\"")});
  EXPECT_EQ(Expected, formatAndApply(Code, Replaces));
```

There is also a test for preserving the byte order mark.

```C++
  verifyFormat("\xef\xbb\xbf\n#include <iostream>");
```

**Review by HazardyKnusperkeks — APPROVED** [review]

**Review by mydeveloperday — APPROVED** [review]

**Comment by owenca** [comment]
I think this is a good time to combine the `KeepEmptyLines...` options. For example:
```
KeepEmptyLines:
  AtStartOfBlock: true
  AtStartOfFile: false
  AtEndOfFile: true
```

**Comment by owenca** [comment]
> Should I make dropping leading blank lines the default?

We should keep leading blank lines (up to `MaxEmptyLinesToKeep`) by default as before.

**Review by owenca — CHANGES_REQUESTED** [review]
See https://github.com/llvm/llvm-project/pull/91221#issuecomment-2105454428.

**Comment by owenca** [comment]
Please refer to e3eca335940251308c8990c8880341002e74b9c1 for grouping/deprecating existing options with a new one.

File: clang/lib/Format/ContinuationIndenter.cpp [hunk 1]
```diff
@@ -208,6 +208,9 @@ RawStringFormatStyleManager::RawStringFormatStyleManager(
       LanguageStyle = PredefinedStyle;
     }
     LanguageStyle->ColumnLimit = CodeStyle.ColumnLimit;
+    // This way the first line of the string does not have to follow the code
+    // before the string.
+    LanguageStyle->KeepEmptyLines.AtStartOfFile = true;
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/ContinuationIndenter.cpp:213]: Why exactly is this necessary?

**Comment by sstwcw** [comment]
I was initially going to make removing leading blank lines the default.
Now the line is not needed anymore.  But I forgot to remove it.

File: clang/unittests/Format/FormatTest.cpp [hunk 2]
```diff
@@ -45,6 +45,9 @@ TEST_F(FormatTest, FormatsGlobalStatementsAt0) {
   verifyFormat("\nint i;", " \n\t \v \f  int i;");
   verifyFormat("int i;\nint j;", "    int i; int j;");
   verifyFormat("int i;\nint j;", "    int i;\n  int j;");
+  auto Style = getLLVMStyle();
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:48]: An empty line before `auto Style = ...`?

File: clang/unittests/Format/FormatTest.cpp [hunk 3]
```diff
@@ -21947,6 +21950,10 @@ TEST_F(FormatTest, HandlesUTF8BOM) {
   verifyFormat("\xef\xbb\xbf");
   verifyFormat("\xef\xbb\xbf#include <iostream>");
   verifyFormat("\xef\xbb\xbf\n#include <iostream>");
+  auto Style = getLLVMStyle();
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:21953]: Ditto.

**Review by owenca — APPROVED** [review]

**Comment by vitalybuka** [comment]
@sstwcw Please don't ignore the presubmit, it detected issue before merging:

Some checks were not successful
2 skipped, 65 successful, 48 failing, and 3 pending checks

**Pull request merged**